zrml/swaps/src/lib.rs

```diff
@@ -1208,10 +1235,121 @@ mod pallet {
+                extra_weight_per_pool,
+            )
+        }
+
+        // Apply a `mutation` to all pools cached for arbitrage (but at most `pool_count` many) and
+        // return the actual weight consumed.
+        pub(crate) fn apply_to_cached_pools<F>(
+            pool_count: u32,
+            mutation: F,
+            max_weight_per_pool: Weight,
+        ) -> Weight
+        where
+            F: Fn(PoolId) -> Result<Weight, DispatchError>,
+        {
+            let mut total_weight = T::WeightInfo::apply_to_cached_pools_noop(pool_count);
+            for (pool_id, _) in PoolsCachedForArbitrage::<T>::drain().take(pool_count as usize) {
+                // The mutation should never fail, but if it does, we just assume we
+                // consumed all the weight and rollback the pool.
+                let _ = with_transaction(|| match mutation(pool_id) {
+                    Err(err) => {
+                        log::warn!(
+                            "Arbitrage unexpectedly failed on pool {:?} with error: {:?}",
+                            pool_id,
+                            err,
+                        );
+                        total_weight = total_weight.saturating_add(max_weight_per_pool);
+                        TransactionOutcome::Rollback(err.into())
+                    }
+                    Ok(weight) => {
+                        total_weight = total_weight.saturating_add(weight);
+                        TransactionOutcome::Commit(Ok(()))
+                    }
+                });
+            }
+            total_weight
+        }
+
+        // Execute arbitrage on a single pool.
+        pub(crate) fn execute_arbitrage(
+            pool_id: PoolId,
+            max_iterations: usize,
+        ) -> Result<Weight, DispatchError> {
+            let pool = Self::pool_by_id(pool_id)?;
+            let pool_account = Self::pool_account_id(&pool_id);
+            let balances = pool
+                .assets
+                .iter()
+                .map(|a| (*a, T::AssetManager::free_balance(*a, &pool_account)))
+                .collect::<BTreeMap<_, _>>();
+            let asset_count = pool.assets.len() as u32;
+            let outcome_tokens = pool.assets.iter().filter(|a| **a != pool.base_asset);
+            let total_spot_price = pool.calc_total_spot_price(&balances)?;
+
+            if total_spot_price > BASE.saturating_add(ARBITRAGE_THRESHOLD) {
+                let (amount, _) =
+                    pool.calc_arbitrage_amount_mint_sell(&balances, max_iterations)?;
+                // We're faking a `buy_complete_sets` operation by transfering to the market prize
+                // pool.
+                T::AssetManager::transfer(
+                    pool.base_asset,
+                    &pool_account,
+                    &T::MarketCommons::market_account(pool.market_id),
+                    amount,
+                )?;
```

Comment:
I am worried here that the remaining ZTG is less than `ExistentialDeposit`, which will result in dust collection. We might want to use `min` between the lowest `ExistentialDeposit` and the `smallest_balance / 2` at https://github.com/zeitgeistpm/zeitgeist/blob/e5c08a8420bc7f1c900f765a38897ec9e5f02f95/zrml/swaps/src/arbitrage.rs#L179-L186

Reply:
Thanks, great observation. Fixed in 8d02820bb9e95e2f7b2c0a6bb5758cd0901034e5.

Testing this took a while because constructing a situation where the base asset would be moved below the minimum balance is only possible in a situation where the bisection method overshoots its target.